### PR TITLE
Darwin stdenv bump LLVM to 11

### PR DESCRIPTION
**Pull request by toonn**
###### Motivation for this change

This bumps the LLVM used in the Darwin stdenv to 11, building on #121055, which made the required updates to the bootstrap-tools first.

Some packages the stdenv depends on needed patching because of the LLVM update so it's likely there are many packages that require compiler flags, mostly to bring the behavior back in line with clang 7, so I'm opening this as a draft to get a hydra jobset.

I'm building `hello` as a test, will report back tomorrow.

### CODE REVIEW

**Comment by happysalada** [comment]
Not specifically related but the two lines begining with the following in darwin stdenv
```
# Workaround for https://openradar.appspot.com/22671534 on 10.11.
```
are not required anymore. It was like the comment says a workaround for 10.11. I've tried on local and they can be safely removed.
If you would rather leave that to another PR, no worries.

**Comment by happysalada** [comment]
One other thing. Would you not make the PR on staging rather than master? The number of packages to be rebuilt would be quite big in this case.

**Comment by toonn** [comment]
@happysalada, I'm targeting master rn on @LnL7's advice. It's so we have a good hydra evaluation to compare to for the custom jobset for this.

**Comment by toonn** [comment]
@LnL7, `hello` built successfully so this is ready for an evaluation.

**Comment by LnL7** [comment]
Hydra eval is running. Note that this will take a while as this jobset has very low build shares, so other builds from master and staging generally get priority.

https://hydra.nixos.org/eval/1677441?compare=1677022#tabs-unfinished

**Comment by prusnak** [comment]
Shouldn't this be targeted at staging, not master?

**Comment by alyssais** [comment]
@prusnak scroll up https://github.com/NixOS/nixpkgs/pull/126411#issuecomment-858242519

File: pkgs/development/compilers/llvm/11/libcxxabi/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/llvm/11/libcxxabi/default.nix:36]:
```suggestion
  cmakeFlags = lib.optionals standalone [
```

File: pkgs/misc/cups/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/default.nix:44]:
```suggestion
  patches = lib.optional stdenv.isDarwin ./0001-TargetConditionals.patch;
```
So we do not forget it that easily.

File: pkgs/development/libraries/portaudio/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/portaudio/default.nix:5]:
```suggestion
  pname = "portaudio";
  version = "190700_20210406";
```

File: pkgs/development/libraries/portaudio/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/portaudio/default.nix:8]:
```suggestion
    url = "http://files.portaudio.com/archives/pa_stable_v${version}.tgz";
```

File: pkgs/development/libraries/fltk/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/fltk/default.nix:10]: This could be converted to a rec.

File: pkgs/misc/cups/default.nix
Comment (toonn) [reply pkgs/misc/cups/default.nix:44]: I can add a comment but it's more accurate when applied depending on the version.

File: pkgs/development/libraries/fltk/default.nix
Comment (toonn) [reply pkgs/development/libraries/fltk/default.nix]: Is there a reason `rec` is preferred?

File: pkgs/development/libraries/portaudio/default.nix
Comment (toonn) [reply pkgs/development/libraries/portaudio/default.nix]: Thanks for spotting the incorrect version!

File: pkgs/development/libraries/portaudio/default.nix
Comment (prusnak) [inline pkgs/development/libraries/portaudio/default.nix:6]: The version should be `190700_20210406` (see the original `src.url` below).

**Review by prusnak — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/portaudio/default.nix
Comment (toonn) [reply pkgs/development/libraries/portaudio/default.nix:6]: Yes, thank you. Sandro caught it a couple minutes ago : )

File: pkgs/misc/cups/default.nix
Comment (SuperSandro2000) [reply pkgs/misc/cups/default.nix:44]: Thats totally true but with this the patch will most likely fail when updating and we do not forget it.

File: pkgs/development/libraries/fltk/default.nix
Comment (SuperSandro2000) [reply pkgs/development/libraries/fltk/default.nix]: It makes it easier to overwrite.

File: pkgs/development/libraries/portaudio/default.nix
Comment (prusnak) [reply pkgs/development/libraries/portaudio/default.nix:6]: I am keeping this conversation unresolved until it has been addressed in the commit.

Comment (toonn) [reply pkgs/development/libraries/portaudio/default.nix:6]: It was already addressed in [this commit](https://github.com/NixOS/nixpkgs/pull/126411/commits/eb3295d44c8940909350383c6f2b2188b6a8e0b5)

Comment (toonn) [reply pkgs/development/libraries/portaudio/default.nix:6]: No, you're right. I managed to mash up the old and new versions again!
Thanks!

Comment (prusnak) [reply pkgs/development/libraries/portaudio/default.nix]: Confirmed that f24c0b3 is what I meant

File: pkgs/development/python-modules/curio/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/curio/default.nix:23]: Please remove the unused fetchpatch from inputs.

File: pkgs/misc/cups/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/default.nix:44]: Does the patch break Linux? If not please apply it unconditionally.

File: pkgs/servers/search/groonga/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/search/groonga/default.nix:24]: ``autoconf automake pkg-config`` go into nativeBuildInput.

File: pkgs/misc/cups/default.nix
Comment (toonn) [reply pkgs/misc/cups/default.nix:44]: I don't have a way to test handy. I thought it might avoid unnecessary rebuilds.

Comment (toonn) [reply pkgs/misc/cups/default.nix:44]: It shouldn't affect it though, since it's guarded by an `ifdef`.

File: pkgs/top-level/all-packages.nix
Comment (angustrau) [inline pkgs/top-level/all-packages.nix:19475]: `darwin.apple_sdk.sdk` doesn't exist on aarch64. See #131791

Comment (toonn) [reply pkgs/top-level/all-packages.nix]: Thanks for pointing this out. I think we're in luck. `libpulseaudio` doesn't actually seem to require the SDK, Libc and the frameworks suffice to build it.

**Comment by r-burns** [comment]
Not sure if this is the right solution, but this fixes the build for qtbase 5.14 / 5.15

```diff
--- a/src/plugins/platforms/cocoa/qcocoawindow.mm
+++ b/src/plugins/platforms/cocoa/qcocoawindow.mm
@@ -1694,8 +1694,10 @@ void QCocoaWindow::setWindowCursor(NSCursor *cursor)
 void QCocoaWindow::registerTouch(bool enable)
 {
     m_registerTouchCount += enable ? 1 : -1;
+    if (@available(macOS 10.12.2, *)) {
     if (enable && m_registerTouchCount == 1)
         m_view.allowedTouchTypes |= NSTouchTypeMaskIndirect;
     else if (m_registerTouchCount == 0)
         m_view.allowedTouchTypes &= ~NSTouchTypeMaskIndirect;
+    }
 }
```

**Comment by r-burns** [comment]
Fyi, I rebased these commits onto master and stdenv still builds.

**Comment by toonn** [comment]
@r-burns, that's good information. I haven't rebased so far just so we have a fixed point to compare the Hydra evaluation to. I do wonder whether the patch for qtbase actually fixes anything or whether it simply hides the problem?

**Comment by r-burns** [comment]
Yeah not sure. The way I have it currently, applications trying to use 10.12.2-only touch support will build successfully and registerTouch will do nothing. Alternatively we could mark the functions themselves as requiring 10.12.2 so that applications calling them fail to build. Probably doesn't make much of a difference, since either way touch apps won't work until the SDK bump.

**Comment by toonn** [comment]
Yeah, you're right. I'll add the patch unless anyone objects. I guess touch support might not even be common on macs. Only iOS devices have touch screens anyway.

**Comment by r-burns** [comment]
Fix for libtiff: https://github.com/nixos/nixpkgs/commit/a43eecc11c9b86ad770c2eba235e635357c187bc

**Comment by toonn** [comment]
Huh, not sure why I didn't run into it. I've had many issues with version files. Thanks!

**Comment by r-burns** [comment]
weird, it looks like sphinx still failed?

https://hydra.nixos.org/build/155876400

```
trying https://github.com/sphinx-doc/sphinx/archive/v4.0.2.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   125  100   125    0     0    838      0 --:--:-- --:--:-- --:--:--   844
100 6008k    0 6008k    0     0  6351k      0 --:--:-- --:--:-- --:--:-- 6351k
unpacking source archive /build/v4.0.2.tar.gz
mv: cannot stat 'tests/roots/test-images/testima'$'\314\210''ge.png': No such file or directory
builder for '/nix/store/ikcqw7ky50c1hcqb8pnkr8bmh3b9w2g8-source.drv' failed with exit code 1
```

**Comment by toonn** [comment]
Forgot some commits that actually fixed the sphinx problem instead of problems caused by the rebase on staging-next. So I had the evaluation aborted and [another one](https://hydra.nixos.org/eval/1713061?compare=1710908) started.

**Review by alyssais — CHANGES_REQUESTED** [review]
Great work!  I'm so happy this is finally being done.

File: pkgs/development/compilers/llvm/11/libcxxabi/default.nix
Comment (toonn) [reply pkgs/development/compilers/llvm/11/libcxxabi/default.nix:4]: I've done this for LLVM 12 and 13 now. Older LLVM posed some challenges so I'd rather skip them. The only place this is in use so far is the Darwin stdenv and we wouldn't use older LLVMs for that anyway, going forward.

I do want to note that this is basically a cargo-cult of the changes for LLVM 7. I did run into this problem [here](https://gist.github.com/toonn/2dfb04d075db4bdaedb0504339a74046#stdenv-bootstrap-stage-3-ends-up-depending-on-llvm-11-libunwind) but I can't find a way to reproduce it. Even the example from the commit making the changes for LLVM 7 (fece3eb2e981b3bd9ea73b34f302ff3aa3d9f81c) doesn't help because I cannot figure out how to generate `<hash>-libc++-<version>`, note the `++` rather than `xx`.

Comment (alyssais) [reply pkgs/development/compilers/llvm/11/libcxxabi/default.nix:4]: > Older LLVM posed some challenges so I'd rather skip them.

That's unfortunate, but the main thing is that it's in the most recent version, so it gets carried forward when 13 is copied to 14, etc, so I think it's okay.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix
Comment (alyssais) [inline pkgs/development/libraries/qt-5/modules/qtwebengine.nix:116]: Should this be `stdenv.isClang`?

**Review by alyssais — COMMENTED** [review]
I think we're so close.  I'm excited.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix
Comment (toonn) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:116]: Yes, it should : )

**Comment by jonringer** [comment]
I think if you target staging, we can merge this, and use the next staging-next to stabilize the changes.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix
Comment (toonn) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix]: I rebased on a more recent staging-next commit in preparation of more evaluations. Hope it doesn't mess up reviews.

File: pkgs/development/compilers/llvm/11/libcxxabi/default.nix
Comment (toonn) [reply pkgs/development/compilers/llvm/11/libcxxabi/default.nix:4]: Talked to @thefloweringash about this. The reason I couldn't generate a path with `libc++` is that [this changed in a refactor](https://github.com/NixOS/nixpkgs/commit/e0d3c9d0317edf7fd8d99e78066a7cdaf1c5da21)...

**Comment by toonn** [comment]
@jonringer, should I go ahead and do that or wait until we've had a thorough review or two?

**Review by alyssais — APPROVED** [review]
As far as I'm concerned this is GTG as long as it's autosquashed and retargeted to staging.

File: pkgs/misc/cups/0001-TargetConditionals.patch
Comment (jonringer) [inline pkgs/misc/cups/0001-TargetConditionals.patch:23]: This should probably be upstreamed.

Wasn't created cups created originally for apple?

**Review by jonringer — APPROVED** [review]
Curious why cups needs to be patched by nixpkgs for macos, when it's under the apple umbrella. But I'm not going to block this PR because it's odd, that just seems to be the current state of things

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix
Comment (jonringer) [inline pkgs/development/libraries/qt-5/modules/qtwebengine.nix:113]:
```suggestion
  ] ++ lib.optionals stdenv.cc.isClang [
```

```
error: attribute 'isClang' missing

       at /home/jon/.cache/nixpkgs-review/pr-126411/nixpkgs/pkgs/development/libraries/qt-5/modules/qtwebengine.nix:113:22:

          112|     "-march=westmere"
          113|   ] ++ lib.optionals stdenv.isClang [
             |                      ^
          114|     "-Wno-elaborated-enum-base"

       … while evaluating 'optionals'
```

File: pkgs/misc/cups/0001-TargetConditionals.patch
Comment (toonn) [reply pkgs/misc/cups/0001-TargetConditionals.patch:23]: Yes, the problem is that we don't use an Xcode build environment, which is probably a project assumption when building for macOS. So I'm not sure this is upstreamable.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix
Comment (toonn) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:113]: Good catch. Should definitely have checked this.

**Review by thiagokokada — APPROVED** [review]
This looks beautiful, congrats :clap: .

**Comment by jonringer** [comment]
Future me, is going to hate current me. But lets get this going.

**Comment by jonringer** [comment]
sorry, trying to build ripgrep on my server, to ensure something isn't fundamentally wrong all the way through to the rust toolchain. give me a few hours to do the stdenv rebuild process :(

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 126411 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>1 package failed to build:</summary>  <ul>    <li>ripgrep</li>  </ul></details>

**Comment by trofi** [comment]
Looks like `mariadb-client` needs a fix:

```
$ nix build -f. --no-link mariadb-client -L
mariadb-client> unpacking sources
mariadb-client> unpacking source archive /nix/store/zk6b3f70aqyxd8pasyvzl9knmwja24cj-mariadb-10.6.5.tar.gz
mariadb-client> source root is mariadb-10.6.5
mariadb-client> setting SOURCE_DATE_EPOCH to timestamp 1636142613 of file mariadb-10.6.5/wsrep-lib/wsrep-API/v26/wsrep_uuid.c
mariadb-client> patching sources
mariadb-client> mv: cannot stat 'storage/mroonga/version': No such file or directory
```

**Comment by toonn** [comment]
@trofi, addressed the mariadb issue in #149096.

**Comment by vcunat** [comment]
[Nix won't build](https://hydra.nixos.org/build/160778459), probably due to this?  There are also other regressions; please have a look at staging-next  https://github.com/NixOS/nixpkgs/pull/148396#issuecomment-991033850

**Comment by happysalada** [comment]
Oh that was the upgrade. I wasn't aware that this was merged actually. I'm curious if other darwin people have an opinion on the nix failure on staging-next. I haven't been able to figure it out.

**Comment by toonn** [comment]
Sounds to me like it's trying to use a method introduced in SDK 10.13 and Nixpkgs is currently on SDK 10.12? I assume Clang 11 is just louder about the error and it doesn't actually cause any problems for anyone who's not on 10.12, hence not having been noticed before?

**Comment by r-burns** [comment]
I think the availability macros in libcxx are only applicable for the system /usr/lib/libc++.dylib. I think anything built purely in nixpkgs should be able to define _LIBCPP_DISABLE_AVAILABILITY without any issues.

File: pkgs/development/python-modules/cffi/default.nix
Comment (bobrik) [reply pkgs/development/python-modules/cffi/default.nix:33]: I think this broke aarch64-darwin:

```
_____________ ERROR collecting testing/cffi1/test_parse_c_type.py ______________
testing/cffi1/test_parse_c_type.py:74: in <module>
    def fetch_constant_five(p):
cffi/api.py:396: in callback_decorator_wrap
    return self._backend.callback(cdecl, python_callable,
E   MemoryError: Cannot allocate write+execute memory for ffi.callback(). You might be running on a system that prevents this. For more information, see https://cffi.readthedocs.io/en/latest/using.html#callbacks
```

Comment (bobrik) [reply pkgs/development/python-modules/cffi/default.nix:33]: https://github.com/NixOS/nixpkgs/pull/152166

**Pull request merged**
